### PR TITLE
[Sol2Yul] Implements function-to-function cast.

### LINKED ISSUE #8349: [Sol2Yul] Implement type conversions
I'm not sure if no type conversions at all are implemented, but at least the following is missing:

[solidity::frontend::YulUtilFunctions::conversionFunctionSpecial](https://github.com/ethereum/solidity/blob/develop/libsolidity/codegen/YulUtilFunctions.cpp#L1884).

### PR DESCRIPTION
**Pull request by christianparpart**
Prerequisite for PR #8246.

Closes: #8349

### CODE REVIEW

**Comment by Marenz** [comment]
please add a test case for this (or enable one of the existing non-yul tests to test yul as well)

File: test/libsolidity/semanticTests/viaYul/conversion/function_cast.sol [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+        return 2 * x;
+    }
+    function g() public view returns (function (uint) external returns (uint)) {
+        return this.f;
+    }
+    function h(uint x) public returns (uint) {
+        return this.g()(x) + 1;
+    }
+}
+// ----
```
Comment (ekpyron) [inline test/libsolidity/semanticTests/viaYul/conversion/function_cast.sol:20]: Might it be reasonable to compile this via Yul, if it is to serve as test for this ;-)?

Comment (ekpyron) [reply test/libsolidity/semanticTests/viaYul/conversion/function_cast.sol:20]: Also you could use an even simpler test for just testing the cast itself like, but I'm actually not sure that's better, but maybe doesn't hurt to just add it as well:
```
  function t() external view returns (function() external a, function() external view b) {
    a = this.f; b = this.f;
  }
```

**Review by ekpyron — CHANGES_REQUESTED** [review]

File: test/libsolidity/semanticTests/viaYul/conversion/function_cast.sol [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+		a = this.f;
+		b = this.f;
+	}
+}
+// ====
+// compileViaYul: also
+// ----
+// f(uint256): 2 -> 4
+// h(uint256): 2 -> 5
+// t() -> 7175878113405833249322534082293024008058894263817453456640262965071783657472, 7175878113405833249322534082293024008058894263817453456640262965071783657472
```
Comment (ekpyron) [inline test/libsolidity/semanticTests/viaYul/conversion/function_cast.sol]: Hexadecimal expectations would be nicer...

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 3]
```diff
@@ -1333,6 +1333,35 @@ string YulUtilFunctions::allocateMemoryArrayFunction(ArrayType const& _type)
 
 string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 {
+	if (_from.category() == Type::Category::Function)
```
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp:1336]: We could still think about extracting this to some function or to have some nice multi-dispatch mechanism for this in general, but I'd say it's fine for now...

**Review by ekpyron — COMMENTED** [review]
Looks good, one minor comment only.

File: test/libsolidity/semanticTests/viaYul/conversion/function_cast.sol (hunk 2 above)
Comment (christianparpart) [reply test/libsolidity/semanticTests/viaYul/conversion/function_cast.sol]: ...

**Review by ekpyron — APPROVED** [review]
Instead of the identity conversion we could have a general mechanism for signifying "does not need conversion" (@christianparpart and I had been talking about that and we weren't sure what's better).
In any case, I think this should work fine, so I'll approve.

**Pull request merged**